distributed/semaphore.py

```diff
@@ -0,0 +1,259 @@
+                # been released and we can try to acquire again (continue loop)
+                if not result:
+                    future = asyncio.wait_for(
+                        self.events[name].wait(), timeout=w.leftover()
+                    )
+                    try:
+                        await future
+                        continue
+                    except TimeoutError:
+                        result = False
+                return result
+
+    async def release(self, comm=None, name=None, client=None, identifier=None):
+        with log_errors():
+            if isinstance(name, list):
+                name = tuple(name)
+            if name in self.leases and identifier in self.leases[name]:
+                self.scheduler.loop.add_callback(
+                    self._release_value, name, client, identifier
+                )
+            else:
+                raise ValueError("Semaphore released too many times.")
+
+    async def _release_value(self, name, client, identifier):
+        self.leases_per_client[client][name].remove(identifier)
+        self.leases[name].remove(identifier)
+        self.events[name].set()
+
+    def _release_client(self, client):
+        semaphore_names = list(self.leases_per_client[client])
+        for name in semaphore_names:
+            ids = list(self.leases_per_client[client][name])
+            for _id in list(ids):
+                self.scheduler.loop.add_callback(
+                    self._release_value, name=name, client=client, identifier=_id
+                )
+
+    async def _validate_leases(self):
+        if not self._validation_running:
+            self._validation_running = True
+            known_clients_with_leases = set(self.leases_per_client.keys())
+            scheduler_clients = set(self.scheduler.clients.keys())
+            for dead_client in known_clients_with_leases - scheduler_clients:
+                client_has_leases = sum(
+                    valmap(len, self.leases_per_client[dead_client]).values()
+                )
+                if client_has_leases:
+                    self._release_client(dead_client)
+            else:
+                self._validation_running = False
+
+
+class Semaphore:
+    def __init__(self, max_leases=1, name=None, client=None):
+        self.client = client or _get_global_client() or get_worker().client
+        self.id = uuid.uuid4().hex
+        self.name = name or "semaphore-" + uuid.uuid4().hex
+        self.max_leases = max_leases
+
+        # TODO: Good idea to create the resource in the init?
+        if self.client.asynchronous:
+            self._started = self.client.scheduler.semaphore_create(
+                name=self.name, max_leases=max_leases
+            )
+        else:
+            self.client.sync(
+                self.client.scheduler.semaphore_create,
+                name=self.name,
+                max_leases=max_leases,
+            )
+            self._started = True
```

Comment:
Maybe this should be an awaitable like `asyncio.sleep(0)` just in case someone awaits the Semaphore later?

(people sometime mix sync and async APIs)